// @atlaskit/calendar/dist/cjs/internal/hooks/use-get-weeks.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useGetWeeks;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _constants = require("../constants");

var _dateToString = _interopRequireDefault(require("../utils/date-to-string"));

var _getBaseCalendar = _interopRequireDefault(require("../utils/get-base-calendar"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function useGetCalendarWithSixthWeek(calendar, _ref) {
  var month = _ref.month,
      year = _ref.year,
      weekStartDay = _ref.weekStartDay;
  var shouldDisplaySixthWeek = calendar.length % 6;
  var calendarLastValue = calendar[calendar.length - 1];
  return (0, _react.useMemo)(function () {
    // Some months jump between 5 and 6 weeks to display. In some cases 4 (Feb
    // with the 1st on a Monday etc). This ensures the UI doesn't jump around by
    // catering to always showing 6 weeks.
    if (shouldDisplaySixthWeek) {
      var lastDayIsSibling = calendarLastValue.siblingMonth;
      var sliceStart = lastDayIsSibling ? _constants.daysPerWeek : 0;
      return (0, _getBaseCalendar.default)(year, month, {
        weekStartDay: weekStartDay
      }).slice(sliceStart, sliceStart + _constants.daysPerWeek).map(function (date) {
        return _objectSpread(_objectSpread({}, date), {}, {
          siblingMonth: true
        });
      });
    }
  }, [calendarLastValue, month, shouldDisplaySixthWeek, weekStartDay, year]);
}

function useGetWeeks(_ref2) {
  var day = _ref2.day,
      month = _ref2.month,
      year = _ref2.year,
      today = _ref2.today,
      disabled = _ref2.disabled,
      selected = _ref2.selected,
      previouslySelected = _ref2.previouslySelected,
      weekStartDay = _ref2.weekStartDay;
  var calendar = (0, _react.useMemo)(function () {
    return (0, _getBaseCalendar.default)(year, month - 1, {
      weekStartDay: weekStartDay
    });
  }, [month, weekStartDay, year]);
  var calendarWithSixthWeek = useGetCalendarWithSixthWeek(calendar, {
    month: month,
    year: year,
    weekStartDay: weekStartDay
  });

  if (calendarWithSixthWeek) {
    calendar.push.apply(calendar, (0, _toConsumableArray2.default)(calendarWithSixthWeek));
  }

  return (0, _react.useMemo)(function () {
    var weeks = [];
    calendar.forEach(function (date) {
      var dateAsString = (0, _dateToString.default)(date, {
        fixMonth: true
      });
      var week;

      if (date.weekDay === weekStartDay) {
        week = {
          id: dateAsString,
          values: []
        };
        weeks.push(week);
      } else {
        week = weeks[weeks.length - 1];
      }

      var isDisabled = disabled.indexOf(dateAsString) > -1;
      var isFocused = day === date.day && !date.siblingMonth;
      var isPreviouslySelected = !isDisabled && previouslySelected.indexOf(dateAsString) > -1;
      var isSelected = !isDisabled && selected.indexOf(dateAsString) > -1;
      var isSiblingMonth = !!date.siblingMonth;
      var isToday = today === dateAsString;
      week.values.push({
        id: dateAsString,
        isDisabled: isDisabled,
        isFocused: isFocused,
        isToday: isToday,
        month: date.month + 1,
        isPreviouslySelected: isPreviouslySelected,
        isSelected: isSelected,
        isSiblingMonth: isSiblingMonth,
        year: date.year,
        day: date.day
      });
    });
    return weeks;
  }, [calendar, day, disabled, previouslySelected, selected, today, weekStartDay]);
}
