// @atlaskit/calendar/dist/cjs/calendar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _core = require("@emotion/core");

var _usePlatformLeafEventHandler = require("@atlaskit/analytics-next/usePlatformLeafEventHandler");

var _noop = _interopRequireDefault(require("@atlaskit/ds-lib/noop"));

var _components = _interopRequireDefault(require("@atlaskit/theme/components"));

var _heading = _interopRequireDefault(require("./internal/components/heading"));

var _weekDays = _interopRequireDefault(require("./internal/components/week-days"));

var _weekHeader = _interopRequireDefault(require("./internal/components/week-header"));

var _constants = require("./internal/constants");

var _useCalendarRef = _interopRequireDefault(require("./internal/hooks/use-calendar-ref"));

var _useControlledDateState = _interopRequireDefault(require("./internal/hooks/use-controlled-date-state"));

var _useFocusing2 = _interopRequireDefault(require("./internal/hooks/use-focusing"));

var _useGetWeeks = _interopRequireDefault(require("./internal/hooks/use-get-weeks"));

var _useHandleDateChange2 = _interopRequireDefault(require("./internal/hooks/use-handle-date-change"));

var _useHandleDateSelect2 = _interopRequireDefault(require("./internal/hooks/use-handle-date-select"));

var _useLocale2 = _interopRequireDefault(require("./internal/hooks/use-locale"));

var _useUniqueId = _interopRequireDefault(require("./internal/hooks/use-unique-id"));

var _container = require("./internal/styles/container");

var _grid = require("./internal/styles/grid");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var analyticsAttributes = {
  componentName: 'calendar',
  packageName: "@atlaskit/calendar",
  packageVersion: "11.1.0"
};
var CalendarWithMode = /*#__PURE__*/(0, _react.forwardRef)(function Calendar(_ref, ref) {
  var _ref$day = _ref.day,
      day = _ref$day === void 0 ? undefined : _ref$day,
      _ref$defaultDay = _ref.defaultDay,
      defaultDay = _ref$defaultDay === void 0 ? 0 : _ref$defaultDay,
      _ref$defaultDisabled = _ref.defaultDisabled,
      defaultDisabled = _ref$defaultDisabled === void 0 ? _constants.blankStringArray : _ref$defaultDisabled,
      _ref$defaultMonth = _ref.defaultMonth,
      defaultMonth = _ref$defaultMonth === void 0 ? 0 : _ref$defaultMonth,
      _ref$defaultPreviousl = _ref.defaultPreviouslySelected,
      defaultPreviouslySelected = _ref$defaultPreviousl === void 0 ? _constants.blankStringArray : _ref$defaultPreviousl,
      _ref$defaultSelected = _ref.defaultSelected,
      defaultSelected = _ref$defaultSelected === void 0 ? _constants.blankStringArray : _ref$defaultSelected,
      _ref$defaultYear = _ref.defaultYear,
      defaultYear = _ref$defaultYear === void 0 ? 0 : _ref$defaultYear,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? undefined : _ref$disabled,
      _ref$month = _ref.month,
      month = _ref$month === void 0 ? undefined : _ref$month,
      _ref$onBlur = _ref.onBlur,
      onBlur = _ref$onBlur === void 0 ? _noop.default : _ref$onBlur,
      _ref$onChange = _ref.onChange,
      onChange = _ref$onChange === void 0 ? _noop.default : _ref$onChange,
      _ref$onFocus = _ref.onFocus,
      onFocus = _ref$onFocus === void 0 ? _noop.default : _ref$onFocus,
      _ref$onSelect = _ref.onSelect,
      onSelect = _ref$onSelect === void 0 ? _noop.default : _ref$onSelect,
      _ref$previouslySelect = _ref.previouslySelected,
      previouslySelected = _ref$previouslySelect === void 0 ? undefined : _ref$previouslySelect,
      _ref$selected = _ref.selected,
      selected = _ref$selected === void 0 ? undefined : _ref$selected,
      _ref$today = _ref.today,
      today = _ref$today === void 0 ? undefined : _ref$today,
      _ref$locale = _ref.locale,
      locale = _ref$locale === void 0 ? 'en-US' : _ref$locale,
      _ref$year = _ref.year,
      year = _ref$year === void 0 ? undefined : _ref$year,
      analyticsContext = _ref.analyticsContext,
      _ref$weekStartDay = _ref.weekStartDay,
      weekStartDay = _ref$weekStartDay === void 0 ? 0 : _ref$weekStartDay,
      testId = _ref.testId,
      calendarRef = _ref.calendarRef,
      mode = _ref.mode,
      className = _ref.className,
      style = _ref.style;

  var _useControlledDateSta = (0, _useControlledDateState.default)({
    day: day,
    defaultDay: defaultDay,
    month: month,
    defaultMonth: defaultMonth,
    year: year,
    defaultYear: defaultYear,
    today: today,
    disabled: disabled,
    defaultDisabled: defaultDisabled,
    selected: selected,
    defaultSelected: defaultSelected,
    previouslySelected: previouslySelected,
    defaultPreviouslySelected: defaultPreviouslySelected
  }),
      _useControlledDateSta2 = (0, _slicedToArray2.default)(_useControlledDateSta.day, 2),
      dayValue = _useControlledDateSta2[0],
      setDayValue = _useControlledDateSta2[1],
      _useControlledDateSta3 = (0, _slicedToArray2.default)(_useControlledDateSta.month, 2),
      monthValue = _useControlledDateSta3[0],
      setMonthValue = _useControlledDateSta3[1],
      _useControlledDateSta4 = (0, _slicedToArray2.default)(_useControlledDateSta.year, 2),
      yearValue = _useControlledDateSta4[0],
      setYearValue = _useControlledDateSta4[1],
      _useControlledDateSta5 = (0, _slicedToArray2.default)(_useControlledDateSta.today, 1),
      todayValue = _useControlledDateSta5[0],
      _useControlledDateSta6 = (0, _slicedToArray2.default)(_useControlledDateSta.disabled, 1),
      disabledValue = _useControlledDateSta6[0],
      _useControlledDateSta7 = (0, _slicedToArray2.default)(_useControlledDateSta.selected, 2),
      selectedValue = _useControlledDateSta7[0],
      setSelectedValue = _useControlledDateSta7[1],
      _useControlledDateSta8 = (0, _slicedToArray2.default)(_useControlledDateSta.previous, 2),
      previouslySelectedValue = _useControlledDateSta8[0],
      setPreviouslySelectedValue = _useControlledDateSta8[1];

  var onChangeWithAnalytics = (0, _usePlatformLeafEventHandler.usePlatformLeafEventHandler)(_objectSpread({
    fn: onChange,
    action: 'changed',
    analyticsData: analyticsContext
  }, analyticsAttributes));

  var _useHandleDateChange = (0, _useHandleDateChange2.default)({
    day: [dayValue, setDayValue],
    month: [monthValue, setMonthValue],
    year: [yearValue, setYearValue],
    onChange: onChangeWithAnalytics
  }),
      navigate = _useHandleDateChange.navigate,
      handleClickNext = _useHandleDateChange.handleClickNext,
      handleClickPrev = _useHandleDateChange.handleClickPrev;

  var onSelectWithAnalytics = (0, _usePlatformLeafEventHandler.usePlatformLeafEventHandler)(_objectSpread({
    fn: onSelect,
    action: 'selected',
    analyticsData: analyticsContext
  }, analyticsAttributes));

  var _useHandleDateSelect = (0, _useHandleDateSelect2.default)({
    day: [dayValue, setDayValue],
    month: [monthValue, setMonthValue],
    year: [yearValue, setYearValue],
    selected: [selectedValue, setSelectedValue],
    previous: [, setPreviouslySelectedValue],
    onSelect: onSelectWithAnalytics,
    navigate: navigate
  }),
      handleClickDay = _useHandleDateSelect.handleClickDay,
      handleContainerKeyDown = _useHandleDateSelect.handleContainerKeyDown;

  var _useFocusing = (0, _useFocusing2.default)({
    onFocus: onFocus,
    onBlur: onBlur
  }),
      handleContainerBlur = _useFocusing.handleContainerBlur,
      handleContainerFocus = _useFocusing.handleContainerFocus;

  (0, _useCalendarRef.default)(calendarRef, {
    navigate: navigate
  });
  var weeks = (0, _useGetWeeks.default)({
    day: dayValue,
    month: monthValue,
    year: yearValue,
    today: todayValue,
    disabled: disabledValue,
    selected: selectedValue,
    previouslySelected: previouslySelectedValue,
    weekStartDay: weekStartDay
  });
  var announceId = (0, _useUniqueId.default)('announce');
  var announcerDate = (0, _react.useMemo)(function () {
    return new Date(yearValue, monthValue - 1, dayValue).toString();
  }, [dayValue, monthValue, yearValue]);

  var _useLocale = (0, _useLocale2.default)({
    locale: locale,
    weekStartDay: weekStartDay
  }),
      monthsLong = _useLocale.monthsLong,
      daysShort = _useLocale.daysShort;

  var wrapperStyle = (0, _react.useMemo)(function () {
    return (0, _container.wrapperStyle)(mode);
  }, [mode]);
  return (0, _core.jsx)("div", {
    className: className,
    style: style,
    onBlur: handleContainerBlur,
    onFocus: handleContainerFocus,
    onKeyDown: handleContainerKeyDown,
    role: "presentation",
    "data-testid": testId && "".concat(testId, "--container"),
    ref: ref
  }, (0, _core.jsx)("div", {
    css: _container.announcerStyle,
    id: announceId,
    "aria-live": "assertive",
    "aria-relevant": "text"
  }, announcerDate), (0, _core.jsx)("div", {
    css: wrapperStyle,
    "aria-describedby": announceId,
    "aria-label": "calendar",
    role: "grid",
    tabIndex: 0
  }, (0, _core.jsx)(_heading.default // The month number needs to be translated to index in the month
  // name array e.g. 1 (January) -> 0
  , {
    monthLongTitle: monthsLong[monthValue - 1],
    year: yearValue,
    handleClickNext: handleClickNext,
    handleClickPrev: handleClickPrev,
    mode: mode,
    testId: testId
  }), (0, _core.jsx)("div", {
    css: _grid.gridsContainerStyle,
    role: "presentation"
  }, (0, _core.jsx)(_weekHeader.default, {
    daysShort: daysShort,
    mode: mode
  }), (0, _core.jsx)(_weekDays.default, {
    weeks: weeks,
    handleClickDay: handleClickDay,
    mode: mode,
    testId: testId
  }))));
});
var Calendar = /*#__PURE__*/(0, _react.memo)( /*#__PURE__*/(0, _react.forwardRef)(function Calendar(props, ref) {
  return (0, _core.jsx)(_components.default.Consumer, null, function (_ref2) {
    var mode = _ref2.mode;
    return (0, _core.jsx)(CalendarWithMode, (0, _extends2.default)({}, props, {
      mode: mode,
      ref: ref
    }));
  });
}));
Calendar.displayName = 'Calendar';
var _default = Calendar;
exports.default = _default;
