// @atlaskit/calendar/dist/es2019/internal/components/week-days.js
/** @jsx jsx */
import { memo } from 'react';
import { jsx } from '@emotion/core';
import { daysGridStyle } from '../styles/grid';
import DateComponent from './date';
const WeekDays = /*#__PURE__*/memo(function WeekDays({
  weeks,
  handleClickDay,
  mode,
  testId
}) {
  return jsx("div", {
    role: "grid",
    "data-testid": testId && `${testId}--month`
  }, weeks.map(week => jsx("div", {
    role: "row",
    key: week.id,
    css: daysGridStyle
  }, week.values.map(({
    id,
    isDisabled,
    isFocused,
    isToday,
    month,
    isPreviouslySelected,
    isSelected,
    isSiblingMonth,
    year,
    day
  }) => jsx(DateComponent, {
    key: id,
    isDisabled: isDisabled,
    isFocused: isFocused,
    isToday: isToday,
    month: month,
    onClick: handleClickDay,
    isPreviouslySelected: isPreviouslySelected,
    isSelected: isSelected,
    isSibling: isSiblingMonth,
    year: year,
    mode: mode,
    testId: testId
  }, day)))));
});
WeekDays.displayName = 'WeekDays';
export default WeekDays;
