// @atlaskit/calendar/dist/cjs/internal/components/heading.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = require("react");

var _core = require("@emotion/core");

var _standardButton = _interopRequireDefault(require("@atlaskit/button/standard-button"));

var _chevronLeftLarge = _interopRequireDefault(require("@atlaskit/icon/glyph/chevron-left-large"));

var _chevronRightLarge = _interopRequireDefault(require("@atlaskit/icon/glyph/chevron-right-large"));

var _colors = require("@atlaskit/theme/colors");

var _heading = require("../styles/heading");

/** @jsx jsx */
var Heading = /*#__PURE__*/(0, _react.memo)(function Heading(_ref) {
  var monthLongTitle = _ref.monthLongTitle,
      year = _ref.year,
      handleClickPrev = _ref.handleClickPrev,
      handleClickNext = _ref.handleClickNext,
      mode = _ref.mode,
      testId = _ref.testId;
  var monthAndYearStyle = (0, _react.useMemo)(function () {
    return (0, _heading.monthAndYearStyle)(mode);
  }, [mode]);
  return (0, _core.jsx)("div", {
    css: _heading.headingStyle,
    "aria-hidden": "true"
  }, (0, _core.jsx)(_standardButton.default, {
    css: _heading.arrowLeftButtonStyle,
    appearance: "subtle",
    spacing: "none",
    tabIndex: -1,
    onClick: handleClickPrev,
    testId: testId && "".concat(testId, "--previous-month"),
    iconBefore: (0, _core.jsx)(_chevronLeftLarge.default, {
      label: "Last month",
      size: "medium",
      primaryColor: _colors.N70
    })
  }), (0, _core.jsx)("div", {
    css: monthAndYearStyle,
    "data-testid": testId && "".concat(testId, "--current-month-year")
  }, "".concat(monthLongTitle, " ").concat(year)), (0, _core.jsx)(_standardButton.default, {
    css: _heading.arrowRightButtonStyle,
    appearance: "subtle",
    spacing: "none",
    tabIndex: -1,
    onClick: handleClickNext,
    testId: testId && "".concat(testId, "--next-month"),
    iconBefore: (0, _core.jsx)(_chevronRightLarge.default, {
      label: "Next month",
      size: "medium",
      primaryColor: _colors.N70
    })
  }));
});
Heading.displayName = 'Heading';
var _default = Heading;
exports.default = _default;
