// @atlaskit/datetime-picker/dist/cjs/components/DateTimePicker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DateTimePickerWithoutAnalytics = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styled = _interopRequireDefault(require("@emotion/styled"));

var _dateFns = require("date-fns");

var _pick = _interopRequireDefault(require("lodash/pick"));

var _analyticsNext = require("@atlaskit/analytics-next");

var _selectClear = _interopRequireDefault(require("@atlaskit/icon/glyph/select-clear"));

var _select = require("@atlaskit/select");

var colors = _interopRequireWildcard(require("@atlaskit/theme/colors"));

var _constants = require("@atlaskit/theme/constants");

var _internal = require("../internal");

var _DatePicker = _interopRequireDefault(require("./DatePicker"));

var _TimePicker = _interopRequireDefault(require("./TimePicker"));

var _templateObject, _templateObject2, _templateObject3, _templateObject4;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var packageName = "@atlaskit/datetime-picker";
var packageVersion = "10.2.1";
/* eslint-disable react/no-unused-prop-types */

var getBorder = function getBorder(_ref) {
  var appearance = _ref.appearance,
      isFocused = _ref.isFocused,
      isInvalid = _ref.isInvalid;
  var color = colors.N20;

  if (appearance === 'subtle') {
    color = 'transparent';
  }

  if (isFocused) {
    color = colors.B100;
  }

  if (isInvalid) {
    color = colors.R400;
  }

  return "border: 2px solid ".concat(color, ";");
};

var getBorderColorHover = function getBorderColorHover(_ref2) {
  var isFocused = _ref2.isFocused,
      isInvalid = _ref2.isInvalid,
      isDisabled = _ref2.isDisabled;
  var color = colors.N30;

  if (isFocused || isDisabled) {
    return "";
  }

  if (isInvalid) {
    color = colors.R400;
  }

  return "border-color: ".concat(color, ";");
};

var getBackgroundColor = function getBackgroundColor(_ref3) {
  var appearance = _ref3.appearance,
      isFocused = _ref3.isFocused;
  var color = colors.N20;

  if (isFocused) {
    color = colors.N0;
  }

  if (appearance === 'subtle') {
    color = 'transparent';
  }

  return "background-color: ".concat(color, ";");
};

var getBackgroundColorHover = function getBackgroundColorHover(_ref4) {
  var isFocused = _ref4.isFocused,
      isInvalid = _ref4.isInvalid,
      isDisabled = _ref4.isDisabled;
  var color = colors.N30;

  if (isFocused || isDisabled) {
    return "";
  }

  if (isInvalid) {
    color = colors.N0;
  }

  return "background-color: ".concat(color, ";");
};

var Flex = _styled.default.div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n  ", "\n  ", "\n  border-radius: ", "px;\n  display: flex;\n  transition: background-color 200ms ease-in-out, border-color 200ms ease-in-out;\n  &:hover {\n    cursor: ", ";\n    ", "\n    ", "\n  }\n"])), getBackgroundColor, getBorder, (0, _constants.borderRadius)(), function (props) {
  return props.isDisabled ? 'default' : 'pointer';
}, getBackgroundColorHover, getBorderColorHover); // Make DatePicker 50% the width of DateTimePicker
// If rendering an icon container, shrink the TimePicker


var DatePickerContainer = _styled.default.div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2.default)(["\n  flex-basis: 50%;\n  flex-grow: 1;\n  flex-shrink: 0;\n"])));

var TimePickerContainer = _styled.default.div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2.default)(["\n  flex-basis: 50%;\n  flex-grow: 1;\n"])));

var ICON_PADDING = 2;

var IconContainer = _styled.default.div(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2.default)(["\n  flex-basis: inherit;\n  padding-left: ", "px;\n  padding-right: ", "px;\n  padding-top: 6px;\n  padding-bottom: 6px;\n  display: flex;\n  align-items: center;\n  color: ", ";\n  transition: color 150ms;\n  &:hover {\n    color: ", ";\n  }\n"])), ICON_PADDING * 2, (0, _constants.gridSize)(), colors.N70, colors.N500); // react-select overrides (via @atlaskit/select).


var styles = {
  control: function control(style) {
    return _objectSpread(_objectSpread({}, style), {}, {
      backgroundColor: 'transparent',
      border: 2,
      borderRadius: 0,
      paddingLeft: 0,
      ':hover': {
        backgroundColor: 'transparent',
        cursor: 'inherit'
      }
    });
  }
};
var dateTimePickerDefaultProps = {
  appearance: 'default',
  autoFocus: false,
  isDisabled: false,
  name: '',
  onBlur: function onBlur(event) {},
  onChange: function onChange(value) {},
  onFocus: function onFocus(event) {},
  innerProps: {},
  id: '',
  defaultValue: '',
  timeIsEditable: false,
  isInvalid: false,
  datePickerProps: {},
  timePickerProps: {},
  datePickerSelectProps: {},
  timePickerSelectProps: {},
  times: _internal.defaultTimes,
  spacing: 'default',
  locale: 'en-US' // Not including a default prop for value as it will
  // Make the component a controlled component

};

var DateTimePicker = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(DateTimePicker, _React$Component);

  var _super = _createSuper(DateTimePicker);

  function DateTimePicker() {
    var _this;

    (0, _classCallCheck2.default)(this, DateTimePicker);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      active: 0,
      dateValue: '',
      isFocused: false,
      timeValue: '',
      value: _this.props.defaultValue,
      zoneValue: ''
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getSafeState", function () {
      var mappedState = _objectSpread(_objectSpread({}, _this.state), (0, _pick.default)(_this.props, ['value']));

      return _objectSpread(_objectSpread({}, mappedState), _this.parseValue(mappedState.value, mappedState.dateValue, mappedState.timeValue, mappedState.zoneValue));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onBlur", function (event) {
      _this.setState({
        isFocused: false
      });

      _this.props.onBlur(event);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onFocus", function (event) {
      _this.setState({
        isFocused: true
      });

      _this.props.onFocus(event);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onDateChange", function (dateValue) {
      _this.onValueChange(_objectSpread(_objectSpread({}, _this.getSafeState()), {}, {
        dateValue: dateValue
      }));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onTimeChange", function (timeValue) {
      _this.onValueChange(_objectSpread(_objectSpread({}, _this.getSafeState()), {}, {
        timeValue: timeValue
      }));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onClear", function () {
      _this.onValueChange(_objectSpread(_objectSpread({}, _this.getSafeState()), {}, {
        timeValue: '',
        dateValue: ''
      }));
    });
    return _this;
  }

  (0, _createClass2.default)(DateTimePicker, [{
    key: "parseValue",
    value: function parseValue(value, dateValue, timeValue, zoneValue) {
      if (this.props.parseValue) {
        return this.props.parseValue(value, dateValue, timeValue, zoneValue);
      }

      var parsed = (0, _dateFns.parse)(value);
      var valid = (0, _dateFns.isValid)(parsed);
      return valid ? {
        dateValue: (0, _dateFns.format)(parsed, 'YYYY-MM-DD'),
        timeValue: (0, _dateFns.format)(parsed, 'HH:mm'),
        zoneValue: (0, _dateFns.format)(parsed, 'ZZ')
      } : {
        dateValue: dateValue,
        timeValue: timeValue,
        zoneValue: zoneValue
      };
    }
  }, {
    key: "onValueChange",
    value: function onValueChange(_ref5) {
      var dateValue = _ref5.dateValue,
          timeValue = _ref5.timeValue,
          zoneValue = _ref5.zoneValue;
      this.setState({
        dateValue: dateValue,
        timeValue: timeValue,
        zoneValue: zoneValue
      });

      if (dateValue && timeValue) {
        var _value = (0, _internal.formatDateTimeZoneIntoIso)(dateValue, timeValue, zoneValue);

        var _this$parseValue = this.parseValue(_value, dateValue, timeValue, zoneValue),
            parsedZone = _this$parseValue.zoneValue;

        var valueWithValidZone = (0, _internal.formatDateTimeZoneIntoIso)(dateValue, timeValue, parsedZone);
        this.setState({
          value: valueWithValidZone
        });
        this.props.onChange(valueWithValidZone); // If the date or time value was cleared when there is an existing datetime value, then clear the value.
      } else if (this.getSafeState().value) {
        this.setState({
          value: ''
        });
        this.props.onChange('');
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          autoFocus = _this$props.autoFocus,
          id = _this$props.id,
          innerProps = _this$props.innerProps,
          isDisabled = _this$props.isDisabled,
          name = _this$props.name,
          timeIsEditable = _this$props.timeIsEditable,
          dateFormat = _this$props.dateFormat,
          datePickerProps = _this$props.datePickerProps,
          datePickerSelectProps = _this$props.datePickerSelectProps,
          timePickerProps = _this$props.timePickerProps,
          timePickerSelectProps = _this$props.timePickerSelectProps,
          times = _this$props.times,
          timeFormat = _this$props.timeFormat,
          locale = _this$props.locale,
          testId = _this$props.testId;

      var _this$getSafeState = this.getSafeState(),
          isFocused = _this$getSafeState.isFocused,
          value = _this$getSafeState.value,
          dateValue = _this$getSafeState.dateValue,
          timeValue = _this$getSafeState.timeValue;

      var bothProps = {
        isDisabled: isDisabled,
        onBlur: this.onBlur,
        onFocus: this.onFocus,
        isInvalid: this.props.isInvalid,
        appearance: this.props.appearance,
        spacing: this.props.spacing
      };
      var _datePickerSelectProp = datePickerSelectProps.styles,
          datePickerStyles = _datePickerSelectProp === void 0 ? {} : _datePickerSelectProp;
      var _timePickerSelectProp = timePickerSelectProps.styles,
          timePickerStyles = _timePickerSelectProp === void 0 ? {} : _timePickerSelectProp;

      var mergedDatePickerSelectProps = _objectSpread(_objectSpread({}, datePickerSelectProps), {}, {
        styles: (0, _select.mergeStyles)(styles, datePickerStyles)
      });

      var mergedTimePickerSelectProps = _objectSpread(_objectSpread({}, timePickerSelectProps), {}, {
        styles: (0, _select.mergeStyles)(styles, timePickerStyles)
      }); // Render DateTimePicker's IconContainer when a value has been filled
      // Don't use Date or TimePicker's because they can't be customised


      var isClearable = Boolean(dateValue || timeValue);
      return /*#__PURE__*/_react.default.createElement(Flex, (0, _extends2.default)({}, innerProps, {
        isFocused: isFocused,
        isDisabled: isDisabled,
        isInvalid: this.props.isInvalid,
        appearance: this.props.appearance
      }), /*#__PURE__*/_react.default.createElement("input", {
        name: name,
        type: "hidden",
        value: value
      }), /*#__PURE__*/_react.default.createElement(DatePickerContainer, null, /*#__PURE__*/_react.default.createElement(_DatePicker.default, (0, _extends2.default)({}, bothProps, {
        autoFocus: autoFocus,
        dateFormat: dateFormat,
        hideIcon: true,
        id: id,
        onChange: this.onDateChange,
        selectProps: mergedDatePickerSelectProps,
        value: dateValue,
        locale: locale,
        testId: testId && "".concat(testId, "--datepicker")
      }, datePickerProps))), /*#__PURE__*/_react.default.createElement(TimePickerContainer, null, /*#__PURE__*/_react.default.createElement(_TimePicker.default, (0, _extends2.default)({}, bothProps, {
        hideIcon: true,
        onChange: this.onTimeChange,
        selectProps: mergedTimePickerSelectProps,
        value: timeValue,
        timeIsEditable: timeIsEditable,
        times: times,
        timeFormat: timeFormat,
        locale: locale,
        testId: testId && "".concat(testId, "--timepicker")
      }, timePickerProps))), isClearable ?
      /*#__PURE__*/
      // eslint-disable-next-line styled-components-a11y/click-events-have-key-events,styled-components-a11y/no-static-element-interactions
      _react.default.createElement(IconContainer, {
        onClick: this.onClear,
        "data-testid": testId && "".concat(testId, "--icon--container")
      }, /*#__PURE__*/_react.default.createElement(_selectClear.default, {
        size: "small",
        primaryColor: "inherit",
        label: "clear"
      })) : null);
    }
  }]);
  return DateTimePicker;
}(_react.default.Component);

exports.DateTimePickerWithoutAnalytics = DateTimePicker;
(0, _defineProperty2.default)(DateTimePicker, "defaultProps", dateTimePickerDefaultProps);

var _default = (0, _analyticsNext.withAnalyticsContext)({
  componentName: 'dateTimePicker',
  packageName: packageName,
  packageVersion: packageVersion
})((0, _analyticsNext.withAnalyticsEvents)({
  onChange: (0, _analyticsNext.createAndFireEvent)('atlaskit')({
    action: 'changed',
    actionSubject: 'dateTimePicker',
    attributes: {
      componentName: 'dateTimePicker',
      packageName: packageName,
      packageVersion: packageVersion
    }
  })
})(DateTimePicker));

exports.default = _default;
